packages/libs/react-ui/src/components/Form/TextField/TextField.tsx

```diff
@@ -1,35 +1,162 @@
+  ChangeEvent,
+  ComponentProps,
+  ElementRef,
+  ForwardedRef,
+  ReactNode,
+} from 'react';
+import React, { forwardRef, useCallback } from 'react';
+import type { AriaTextFieldProps } from 'react-aria';
+import { useFocusRing, useHover, useTextField } from 'react-aria';
+import { formField, inputContainer } from '../Form.css';
 import { FormFieldHeader, FormFieldHelper } from '../FormFieldWrapper';
-import { statusVariant } from '../FormFieldWrapper/FormFieldWrapper.css';
-
-export interface ITextFieldProps
-  extends Omit<IFormFieldWrapperProps, 'children' | 'htmlFor'>,
-    Omit<IInputProps, 'disabled' | 'children'> {}
-
-export const TextField: FC<ITextFieldProps> = forwardRef<
-  HTMLInputElement,
-  ITextFieldProps
->(function TextField(
-  { disabled = false, status, id, label, info, tag, helperText, ...inputProps },
-  ref,
+import { endAddon, input, startAddon } from './TextField.css';
+
+type PickedAriaTextFieldProps = Omit<
+  AriaTextFieldProps,
+  'children' | 'inputElementType' | 'onChange'
+>;
+export interface ITextFieldProps extends PickedAriaTextFieldProps {
+  className?: string;
+  isPositive?: boolean;
+  tag?: string;
+  info?: string;
+  /*
+   * @deprecated Use `isDisabled` instead. only here to support libs that manages props like `react-hook-form`
+   */
+  disabled?: boolean;
+  /*
+   * using native onChange handler instead of AriaTextFieldProps.onChange to support uncontrolled form state eg. react-hook-form
+   */
+  onChange?: ComponentProps<'input'>['onChange'];
+  /*
+   * alias for `AriaTextFieldProps.onChange`
+   */
+  onValueChange?: (value: string) => void;
+  startAddon?: ReactNode;
+  endAddon?: ReactNode;
+}
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type, react/function-component-definition
+export function TextFieldBase(
+  props: ITextFieldProps,
+  forwardedRef: ForwardedRef<ElementRef<'input'>>,
 ) {
-  const statusVal = disabled === true ? 'disabled' : status;
+  const ref = useObjectRef<ElementRef<'input'>>(forwardedRef);
+  const isDisabled = props.isDisabled || props.disabled;
+  const {
+    labelProps,
+    inputProps,
+    descriptionProps,
+    errorMessageProps,
+    ...validation
+  } = useTextField(
+    {
+      ...props,
+      onChange: props.onValueChange,
+      inputElementType: 'input',
+      isDisabled,
+    },
+    ref,
+  );
+
+  const { hoverProps, isHovered } = useHover(props);
+  const { isFocused, isFocusVisible, focusProps } = useFocusRing({
+    isTextInput: true,
+    autoFocus: props.autoFocus,
+  });
+
+  // handle uncontrollable form state eg. react-hook-form
+  const handleOnChange = useCallback(
+    (event: ChangeEvent<ElementRef<'input'>>) => {
+      inputProps.onChange?.(event);
+      props.onChange?.(event);
+    },
+    [props.onChange, inputProps.onChange],
+  );
+
+  // aggregate error message from validation props
+  const errorMessage =
+    typeof props.errorMessage === 'function'
+      ? props.errorMessage(validation)
```

Comment:
in what situation would the `errorMessage` be a function?

Reply:
to customize the native errors I already added story for that

Reply:
https://github.com/kadena-community/kadena.js/pull/1435/files#diff-a6c861377b53963c4c8c0cf5cc655bb99a72802b603ae5f173165627c163f48fR336